Added op-equals and comparison operators

BartMassey opened the pull request:
For my current project I needed to be able to do `<=` and wanted to be able to do `+=` and `-=` and the other comparison operators. These commits provide that functionality.

If you could accept this pull request and then update Cargo I would be grateful. Thanks!

BartMassey committed:
added README body

BartMassey committed:
added += and -=

BartMassey committed:
added comparisons

BartMassey committed:
bumped version

jmitchell commented on src/multiset.rs:
```diff
@@ -298,6 +327,193 @@ impl<T> Sub for HashMultiSet<T> where
+            };
+        };
+        true
+    }
```
This doesn't seem strict enough. For example, it can return true when the right set has elements which aren't present in the left set.

Iterating over the distinct elements of the union of `self` and `other` is a possible fix. Another is applying the same check with `self` and `other` swapped.

BartMassey replied:
Yeah, that's buggy. Should have thought it through more clearly. Nice catch. I'll fix it and revise it.

jmitchell commented on src/multiset.rs:
```diff
@@ -298,6 +327,193 @@ impl<T> Sub for HashMultiSet<T> where
+        };
+    };
+    result
+}
```
What does `lcr` stand for? My gut reaction is this is such a specialized comparison operation that maybe it should be in its own crate or module. It seems a bit strange to impose a particular total ordering on multisets.

BartMassey replied:
`lcr` is "left contains right". I'll give it a better name.

As far as I know, this is the only reasonable partial order on multisets that doesn't depend on the type or ordering of the elements. It is the multiset generalization of the subset partial order on sets, which has the same property. The fact that `HashSet` doesn't support this is either an oversight or a hint that somebody agreed with you. Python, to pick one example, treats comparison operators on sets in this style.

If you feel strongly that co-opting the comparison operators is bad here, I can give the operations some kind of verbose name. `is_subcollection`, `is_proper_subcollection` and so forth are possibilities. Note that `HashSet` gives no test for proper subset, which suggests oversight rather than design to me: this is a thing that programs commonly need to know. I can put these in their own trait (I think?) and/or module in either operator or named form if you like. I don't see any sensible reason to put them in their own crate: they are definitely Multiset-specific and Multiset-useful.

We probably still want '==' and '!=' for equality and inequality. I'll make them total when the elements obey `Eq`, I think. I'll need to think about how to restructure this.

BartMassey replied:
Actually, because HashMap requires Eq on keys, I just gave up and made `==` and `!=` be total equality/inequality.

BartMassey committed:
fixed bug in PartialEq::eq, added Eq, renamed lcr

Thanks to jmitchell for catching the bug in review.

BartMassey committed:
reimplemented subset and proper_subset as named methods

BartMassey commented:
Fixed the paper-bag `eq()` bug, moved the comparison operations to `is_submultiset()` and `is_proper_submultiset()`, and kept `partial_cmp()` as a named function that could be used to quickly implement the inequality operators if desired. Hopefully this is good enough now. I can refactor the commits if you insist, but I think these are reasonably clean?

The app I've written with it, by the way, is <http://github.com/BartMassey/sugarmantra>. Don't know if I mentioned that before.

Thanks for your time on this!

BartMassey committed:
allowed the RHS of +, -, +=, -= to be a ref

BartMassey committed:
bumped version increment for ref changes

BartMassey committed:
generalized Add and Sub a bit more

mashedcode commented:
I don't mind the `PartialEq` and `Sub` implementation. Maybe even `Add` instead of `union` is not a bad idea since this is a multiset. But the Assign Implementations seem weird to me.
Either ways this PR seems stale and needs to be rebased so I'd rather close this for now until someone actively works on this PR again.

jmitchell commented:
I'm closing this since, as @mashedcode points out, it has gone stale. If there's still interest in adding features along these lines, let's discuss them more in the issues.
